waspc/cli/src/Wasp/Cli/GithubRepo.hs

```diff
@@ -0,0 +1,129 @@
+import Data.List (intercalate)
+import Data.Maybe (fromJust, maybeToList)
+import Network.HTTP.Conduit (simpleHttp)
+import qualified Network.HTTP.Simple as HTTP
+import Path.IO (copyDirRecur, removeDirRecur)
+import StrongPath (Abs, Dir, File, Path', (</>))
+import qualified StrongPath as SP
+import StrongPath.Path (toPathAbsDir)
+import System.Directory (createDirectoryIfMissing)
+import Wasp.Cli.FileSystem (getUserCacheDir)
+
+type GithubRepo = (GithubRepoOwner, GithubRepoName)
+
+type GithubRepoOwner = String
+
+type GithubRepoName = String
+
+fetchFolderFromGithubRepoToDisk ::
+  forall d.
+  GithubRepo ->
+  String ->
+  Path' Abs (Dir d) ->
+  IO (Either String ())
+fetchFolderFromGithubRepoToDisk (githubOrg, repoName) pathToFolderInRepo destinationOnDisk = do
+  try
+    (withTempDir ".wasp-templates-temp" downloadFolderToDisk)
+    <&> either showException Right
+  where
+    downloadFolderToDisk :: Path' Abs (Dir d) -> IO ()
+    downloadFolderToDisk tempDir = do
+      downloadFile githubArchiveUrl tempDir archiveFileName >>= unpackArchive tempDir
+      copyRepoFolderToDestination tempDir pathToFolderInRepo destinationOnDisk
+      where
+        githubArchiveUrl = intercalate "/" ["https://github.com", githubOrg, repoName, "archive", archiveFileName]
+        archiveFileName = repoReference ++ ".tar.gz"
+
+    -- Which point in repo history to download (a branch or commit hash)
+    repoReference :: String
+    repoReference = "main"
+
+    downloadFile :: String -> Path' Abs (Dir d) -> String -> IO (Path' Abs (File f))
+    downloadFile downloadUrl destinationDir fileName = do
+      let destinationPath = destinationDir </> fromJust (SP.parseRelFile fileName)
+      simpleHttp downloadUrl >>= BL.writeFile (SP.fromAbsFile destinationPath)
+      return destinationPath
+
+    unpackArchive :: Path' Abs (Dir d) -> Path' Abs (File f) -> IO ()
+    unpackArchive destinationDir pathToTheTarFile = do
+      Tar.unpack (SP.fromAbsDir destinationDir) . Tar.read . GZip.decompress =<< BL.readFile (SP.fromAbsFile pathToTheTarFile)
+
+    copyRepoFolderToDestination :: Path' Abs (Dir d) -> String -> Path' Abs (Dir d) -> IO ()
+    copyRepoFolderToDestination archiveRootDir pathToFolderInRepo' destinationOnDisk' = do
+      copyDirRecur (toPathAbsDir pathToDownloadededFolder) (toPathAbsDir destinationOnDisk')
+      where
+        pathToDownloadededFolder = archiveRootDir </> rootArchiveFolder </> fromJust (SP.parseRelDir pathToFolderInRepo')
+        -- Github repo tars have a root folder that is named after the repo name and the reference (branch or tag).
+        rootArchiveFolder = fromJust (SP.parseRelDir $ repoName ++ "-" ++ repoReference)
+
+    withTempDir :: FilePath -> (Path' Abs (Dir d) -> IO a) -> IO a
+    withTempDir dirName action = do
+      tempDir <- getUserCacheDir <&> (</> fromJust (SP.parseRelDir dirName))
+
+      createDirectoryIfMissing True $ SP.fromAbsDir tempDir
+      result <- action tempDir
```

Comment:
It can be tricky if exception is thrown here, you will never do clean up then.
you should catch it and in "finally" make sure you do cleanup (removing tmp dir).
There is no finally in Haskell, but there is a family of functions called `bracket` that play that role, so use that. Search through codebase for the usage example, and you will probably want to use the same import for it.

Reply:
We agreed to use a thrid-party lib that handles this 👍